presto-parquet/src/main/java/io/prestosql/parquet/reader/TimestampColumnReader.java

```diff
@@ -40,13 +47,22 @@ public TimestampColumnReader(RichColumnDescriptor descriptor, DateTimeZone timeZ
     protected void readValue(BlockBuilder blockBuilder, Type type)
     {
         if (definitionLevel == columnDescriptor.getMaxDefinitionLevel()) {
-            long utcMillis = getTimestampMillis(valuesReader.readBytes());
             if (type instanceof TimestampWithTimeZoneType) {
-                type.writeLong(blockBuilder, packDateTimeWithZone(utcMillis, UTC_KEY));
+                int precision = ((TimestampWithTimeZoneType) type).getPrecision();
+                DecodedTimestamp decodedTimestamp = decode(valuesReader.readBytes());
+                if (precision <= TimestampWithTimeZoneType.MAX_SHORT_PRECISION) {
```

Comment:
We could do a refactor where the type is provided at the construction time (it's const).

(I think i used to have this somewhere...)

add a TODO comment & respective issues